tt_metal/impl/dispatch/kernels/command_queue_producer.hpp

```diff
@@ -109,7 +108,15 @@ void produce(
             // These APIs are non-blocking to allow for context switching.
             if (cb_producer_space_available(num_to_read) and num_reads_issued < num_pages) {
                 uint32_t l1_write_ptr = get_write_ptr(0);
-                src_buffer.noc_async_read_buffer(l1_write_ptr, num_reads_issued, num_to_read, 0);
+                if((BufferType)src_buf_type == BufferType::SYSTEM_MEMORY || !sharded){
+                    Buffer src_buffer((BufferType)src_buf_type, bank_base_address, page_size);
+                    src_buffer.noc_async_read_buffer(l1_write_ptr, num_reads_issued, num_to_read, 0);
+                }
+                else{
+                    // if we are reading from more than one core
+                    ShardedBuffer src_buffer(page_size, num_cores, bank_base_address, command_ptr+6);
+                    src_buffer.noc_async_read_buffer(l1_write_ptr, num_reads_issued, num_to_read);
```

Comment:
see comment above about possible solutions to this non-performant pattern of creating a buffer in the inner loop.